File: bin/heatmap.R
#!/usr/bin/env Rscript
library(tidyverse)
library(optparse)

option_list = list(
  make_option(c("--tsv"), type='character', default = NULL, help="Output from SUMMARIZING_CLUSTERS process", metavar='character'),
  make_option(c("--output"), type='character', default = NULL, help="Output name - without extention", metavar='character')
)
opt_parser <- OptionParser(option_list = option_list)
opt <- parse_args(opt_parser)


# data = read_tsv("/nfs/users/nfs_b/be5/scratch/testing_nextflow/5main_results/jf_extracted100_nrLitMari2_T_extracted100_k21_LC150_minCov150_clustered_nr0_clustered_putativeSLs.tsv")
data = read_tsv(opt$tsv)
numeric_cols <- c("entropy", "avg_coverage", "supporting_kmers", "`#_transcripts_in_cluster`")

normalize_score = function(x) {
  (x -mean(x, na.rm = TRUE))
}

normalize_data = data %>% 
  mutate(across(where(is.numeric), normalize_score, .names = "normalized_{.col}"))

heatmap_data = normalize_data %>%
  select(SL_ID, starts_with("normalized_")) %>%
  pivot_longer(cols = starts_with("normalized_"), names_to = "Metric", values_to = "Normalized_Value") %>%
  mutate(
    Metric = str_replace(Metric, "normalized_", ""),
    Metric = str_replace(Metric, "`#_transcripts_in_cluster`", "#_transcripts_in_cluster"),
    Metric = factor(Metric, levels = c("entropy", "avg_coverage", "supporting_kmers", "#_transcripts_in_cluster")))

heatmap_plot = ggplot(heatmap_data, aes(x = SL_ID, y = Metric, fill = Normalized_Value)) +
                        geom_tile() + 
                        scale_fill_gradient2(
                          low = "blue",
                          mid = "white",
                          high = "red", 
                          midpoint = 0,
                          name = "Normalized value") +
                        theme_minimal() + 
                        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 8))

output = opt$output + "_heatmap.png"
ggsave(output, heatmap_plot, dpi = 600)                      
# heatmap_plot

File: bin/heatmap.py
#!/usr/bin/env python3
import pandas as pd
from plotnine import *
import argparse

parser = argparse.ArgumentParser()
parser.add_argument("--tsv", required=True, help="Input TSV")
parser.add_argument("--output", required=True, help="Output name without extension")
args = parser.parse_args()

# Read data
data = pd.read_csv(args.tsv, sep="\t")

# Columns to normalize
numeric_cols = ["entropy", "avg_coverage", "supporting_kmers", "#_transcripts_in_cluster"]

# numeric_cols <- c("entropy", "avg_coverage", "supporting_kmers", "`#_transcripts_in_cluster`")

# Normalization function
def normalize(x):
    return (x - x.mean())

# Create normalized columns
for col in numeric_cols:
    data[f"normalized_{col}"] = normalize(data[col])

# Reshape to long format
heatmap_data = (
    data.melt(
        id_vars="SL_ID",
        value_vars=[f"normalized_{c}" for c in numeric_cols],
        var_name="Metric",
        value_name="Normalized_Value"
    )
)

# Clean metric names
heatmap_data["Metric"] = heatmap_data["Metric"].str.replace("normalized_", "")

# Order metrics like in R
metric_order = ["entropy", "avg_coverage", "supporting_kmers", "#_transcripts_in_cluster"]
heatmap_data["Metric"] = pd.Categorical(heatmap_data["Metric"], categories=metric_order, ordered=True)

# Build heatmap plot
p = (
    ggplot(heatmap_data, aes("SL_ID", "Metric", fill="Normalized_Value"))
    + geom_tile()
    + scale_fill_gradient2(low="blue", mid="white", high="red", midpoint=0, name="Normalized value")
    + theme_minimal()
    + theme(
        axis_text_x=element_text(angle=90, hjust=1, vjust=0.5, size=8)
    )
)

# Save image
out_file = args.output + "_heatmap.png"
p.save(out_file, dpi=300)
print(f"Saved heatmap to {out_file}")

